src/platforms/common/swdptap.c

```diff
@@ -150,27 +160,26 @@ static void swdptap_seq_out_clk_delay(uint32_t tms_states, size_t clock_cycles)
 static void swdptap_seq_out_clk_delay(const uint32_t tms_states, const size_t clock_cycles)
 {
 	for (size_t cycle = 0; cycle < clock_cycles; ++cycle) {
-		gpio_clear(SWCLK_PORT, SWCLK_PIN);
 		gpio_set_val(SWDIO_PORT, SWDIO_PIN, tms_states & (1 << cycle));
 		for (volatile uint32_t counter = target_clk_divider; counter > 0; --counter)
 			continue;
 		gpio_set(SWCLK_PORT, SWCLK_PIN);
 		for (volatile uint32_t counter = target_clk_divider; counter > 0; --counter)
 			continue;
+		gpio_clear(SWCLK_PORT, SWCLK_PIN);
 	}
-	gpio_clear(SWCLK_PORT, SWCLK_PIN);
 }
 
 static void swdptap_seq_out_no_delay(uint32_t tms_states, size_t clock_cycles) __attribute__((optimize(3)));
 
 static void swdptap_seq_out_no_delay(const uint32_t tms_states, const size_t clock_cycles)
 {
 	for (size_t cycle = 0; cycle < clock_cycles; ++cycle) {
-		gpio_clear(SWCLK_PORT, SWCLK_PIN);
 		gpio_set_val(SWDIO_PORT, SWDIO_PIN, tms_states & (1 << cycle));
 		gpio_set(SWCLK_PORT, SWCLK_PIN);
+		__asm__("nop");
```

Comment:
This 0xbf00 architectural NOP may be consumed by Cortex-M3/M4 in 0 HCLK cycles. Multiple nops can stall the ART, or L1 I-cache, or FLITF prefetch on I-bus with low predictability. If you want to stretch the SWCLK high phase, consider using something different, like a second GPIO write (`gpio_set(SWCLK_PORT, SWCLK_PIN)` again, it should be hot/cached). This use of double writes seems much more justified to me (as opposed to forcing every gpio write be done twice for "faster" stm32f4). Or just "call"/inline my 3(4?)-cycle loop once, whatever is quicker.

Reply:
Due to how this is used, this makes a noticeable difference in the bitbanged output, at least on a STM32F103/GD32F103. It may need different handling on a souped up part, but is correct/fine on `native`.